### PR TITLE
Add support for reply affinity.

### PR DESCRIPTION
**Pull request by smn**
Currently Casepro assumes a single outbound delivery channel.
Junebug supports the concept of channels and messages can arrive at Casepro from
one of many channels.

When we receive a message via Junebug, replies should be routed back to the 
same channel the original message was received on in order to preserve reply affinity.

Each backend will likely have different bits of data that need to be preserved
per message received. This pull-request adds a `metadata` JSON blob in which the 
backend implementation is free to stash whatever information is required to allow
backend specific details to be implemented.

### CODE REVIEW

**Comment by smn** [comment]
@rudigiesler @erikh360 @KaitCrawford @miltontony ready for review.

File: casepro/msgs/models.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 from dash.orgs.models import Org
 from dash.utils import get_obj_cacheable
 from django.contrib.auth.models import User
+from django.contrib.postgres.fields import HStoreField
```
Comment (smn) [inline casepro/msgs/models.py]: Initially used `JSONField` but that relies on postgresql 9.4 as Django (rightly) uses `jsonb` for the underlying schema definition, `jsonb` was only added in 9.4. Using `HStore` as travis is configured for postgresql 9.3 and so I'm assuming some production systems have that as a requirement.

File: casepro/settings_common.py [hunk 2]
```diff
@@ -462,7 +465,10 @@
 LOGIN_REDIRECT_URL = "/"
 LOGOUT_REDIRECT_URL = "/"
 
-AUTHENTICATION_BACKENDS = ('smartmin.backends.CaseInsensitiveBackend',)
+AUTHENTICATION_BACKENDS = (
+    'smartmin.backends.CaseInsensitiveBackend',
+    'guardian.backends.ObjectPermissionBackend',
```
Comment (smn) [inline casepro/settings_common.py]: `./manage.py check` tests were failing on Travis without this. Perhaps a dependency changed to make this required?

File: casepro/backend/junebug.py [hunk 3]
```diff
@@ -273,8 +277,13 @@ class JunebugBackend(BaseBackend):
     def __init__(self):
         self.identity_store = IdentityStore(
             settings.IDENTITY_API_ROOT, settings.IDENTITY_AUTH_TOKEN, settings.IDENTITY_ADDRESS_TYPE)
-        self.message_sender = JunebugMessageSender(
-            settings.JUNEBUG_API_ROOT, settings.JUNEBUG_CHANNEL_ID, settings.JUNEBUG_FROM_ADDRESS, self.identity_store)
+        self.message_senders = dict([
+            (channel_id, JunebugMessageSender(
```
Comment (erikh360) [inline casepro/backend/junebug.py]: We might want to add some validation here, `JUNEBUG_CHANNELS` might be misconfigured.

File: casepro/backend/junebug.py [hunk 4]
```diff
@@ -357,7 +366,22 @@ def push_outgoing(self, org, outgoing, as_broadcast=False):
         :param as_broadcast: whether outgoing messages differ only by recipient and so can be sent as single broadcast
         """
         for message in outgoing:
-            self.message_sender.send_message(message)
+            self.send_message(message)
+
+    def send_message(self, outgoing):
+        message = outgoing.reply_to
```
Comment (erikh360) [inline casepro/backend/junebug.py]: It might be a good idea to add a default channel field on the contact, so we default back to the last channel used by the contact. Not necessary but would be nice.

**Comment by miltontony** [comment]
🍰 from me

File: casepro/backend/junebug.py (hunk 4 above)
Comment (smn) [inline casepro/backend/junebug.py]: Yeah, agreed but I would suggest we do that in a separate PR as it's essentially a different feature and would likely also require UI work to expose that.

File: casepro/backend/junebug.py (hunk 3 above)
Comment (smn) [inline casepro/backend/junebug.py]: good one, I'll add some validation checks to the backend's __init__

Comment (smn) [inline casepro/backend/junebug.py]: @erikh360 validation checks added, please review again?

**Comment by smn** [comment]
tests fixed & settings validation checks introduced for backends. Ready for review.
